### DISCUSSION TITLE
Weird "spotty" adjoint gradients

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2416] [General]
Something interesting that I came across in a few adjoint optimizations recently is the presence of some very tiny "spots" in the gradient. I plotted the gradients returned by the OptimizationProblem's call function and then did the tensorjacobianproduct propagation through the mapping function. The results are interesting.

![image](https://user-images.githubusercontent.com/37753156/221707495-b5fb1958-fca4-4a26-8f86-87b0bcc145ad.png)

The contourplot is of the absolute value of the gradients on a log scale to better see the small gradients so they aren't obscured by the very large gradients at the "spots" which are a few orders of magnitude larger. I thought that it would mess up my optimization, but after a few iterations they disappeared and got smoothed out. This is the geometry that produced the gradients.

![image](https://user-images.githubusercontent.com/37753156/221708469-b5f30898-c557-4f79-ba0d-26553eaefc61.png)

It is obvious that the non-phyiscal, very tiny features are producing the gradients, but I'm not sure why the optimizer decided to change the parameters to that in the first place. After a few iterations the topology got smoothed out.

![image](https://user-images.githubusercontent.com/37753156/221708587-b86bd1ea-f334-4e55-a41c-61d3731b9150.png)

I just thought it was interesting and was wondering if anybody else has encountered this before?

**Comment by smartalecH** [comment]
Try setting `do_averaging=False` within your `MaterialGrid`. The feature is still buggy.  

Edit: the reason this shows up early in the optimization is because typically we use uniform initial conditions. The subpixel smoothing scheme computes an analytic spatial gradient that is very sensitive to numerical artifacts (which will happen if the gradient is zero for a uniform region). It's a known issue, we just need time/resources to address it. 

So you can still use averaging but maybe start with a non uniform initial condition. 

(I'm looking forward to seeing this pol demux GC once it's done)

**Reply by Andeloth** [reply]
Thanks, I'll try some non-uniform initial conditions, and also try `do_averaging=false` a few times.

Haha I'll let you know once it's done :P.
